proxy/src/opentsdb/types.rs

```diff
@@ -0,0 +1,191 @@
+///     http://opentsdb.net/docs/build/html/api_http/put.html#requests
+///
+/// NOTE:
+///     - all the params is unimplemented.
+#[derive(Debug, Default, Deserialize)]
+#[serde(default)]
+pub struct PutParams {
+    // TODO: How to represent a `Present` param ?
+    // now use bool to represent it, so the URL must be "?summary=true/false"
+    // but the `Present` datatype means "?summary" is ok.
+    pub summary: bool,
+    pub details: bool,
+    pub sync: bool,
+    pub sync_timeout: i32,
+}
+
+#[derive(Debug, Deserialize)]
+pub struct Point {
+    pub metric: String,
+    pub timestamp: u64,
+    // TODO: OpenTSDB's value support Integer, Float, String. How to represent it and parse from json ?
+    pub value: f64,
+    pub tags: HashMap<String, String>,
+}
+
+pub(crate) fn convert_put_request(req: PutRequest) -> Result<Vec<WriteTableRequest>> {
+    let points = {
+        // multi points represent as json array
+        let parse_array = from_slice::<Vec<Point>>(&req.points);
+        match parse_array {
+            Ok(points) => Ok(points),
+            Err(_e) => {
+                // single point represent as json object
+                let parse_object = from_slice::<Point>(&req.points);
+                match parse_object {
+                    Ok(point) => Ok(vec![point]),
+                    Err(e) => Err(e),
+                }
+            }
+        }
+    };
+    let points = points.box_err().with_context(|| Internal {
+        msg: "json parse error",
+    })?;
+    validate(&points)?;
+
+    let mut points_per_metric = HashMap::new();
+    for point in points.into_iter() {
+        points_per_metric
+            .entry(point.metric.clone()) // TODO: how to reduce string copy ?
+            .or_insert(Vec::new())
+            .push(point);
+    }
+
+    let requests = points_per_metric
+        .into_iter()
+        .map(|(metric, points)| {
+            let mut tag_names_set = HashSet::new();
+            for point in points.iter() {
+                for tag_name in point.tags.keys() {
+                    tag_names_set.insert(tag_name.clone()); // TODO: how to reduce string copy ?
+                }
+            }
+
+            let mut tag_name_to_tag_index: HashMap<String, u32> = HashMap::new();
+            let mut tag_names = Vec::with_capacity(tag_names_set.len());
+            for (idx, tag_name) in tag_names_set.into_iter().enumerate() {
+                tag_name_to_tag_index.insert(tag_name.clone(), idx as u32);
+                tag_names.push(tag_name);
+            }
+
+            let mut req = WriteTableRequest {
+                table: metric,
+                tag_names,
+                field_names: vec![String::from(OPENTSDB_DEFAULT_FIELD)],
+                entries: Vec::with_capacity(points.len()),
+            };
+
+            for point in points {
+                let mut timestamp = point.timestamp;
+                // TODO: Is there a util function to detect timestamp precision and convert it?
```

Comment:
You can refer to this doc
- https://help.aliyun.com/document_detail/60683.html

Reply:
fixed.